docs/madr/decisions/011-rate-limit-policy.md

```diff
@@ -0,0 +1,294 @@
+ 
+#### Configuration
+ 
+```yaml
+ default:
+   http:
+     requests: 5
+     interval: 10s
+     onRateLimit:
+       status: 423
+       headers:
+         - key: "x-kuma-rate-limited"
+           value: "true"
+           append: true
+```
+ 
+#### **TCP Connection Rate Limit**
+ 
+Rate limiting on TCP connection is more complicated because there is no easy way to match only for specific clients.
+Based on alternatives we decided to make it only available for all the clients.
+ 
+#### Top level
+ 
+```yaml
+targetRef:
+ kind: Mesh|MeshSubset|MeshService|MeshServiceSubset
+ name: ...
+```
+Matching on MeshGatewayRoute and MeshHTTPRoute does not make sense because we are configuring TCP listeners.
+Rate limiting is an inbound policy so only `from` should be configured.
+ 
+#### From level
+ 
+```yaml
+from:
+ - targetRef:
+     kind: Mesh
+     name: ...
+```
+ 
+There is no easy way to match requesting services with the specific filter chain that's why we decided to support only configuring TCP rate limiting from all services.
+ 
+#### Configuration
+ 
+```yaml
+ default:
+   tcp:
+     connections: 100
+     interval: 10s
+```
+ 
+#### **Result**
+ 
+```yaml
+type: MeshRateLimit 
+mesh: example 
+name: example-rate-limit
+spec:
+ targetRef:
+   kind: Mesh|MeshSubset|MeshService|MeshServiceSubset|MeshGatewayRoute|MeshHTTPRoute # or Mesh|MeshSubset|MeshService|MeshServiceSubset for TCP configuration
+   name: example-rate-limit
+ from:
+   - targetRef:
+       kind: Mesh|MeshSubset|MeshService|MeshServiceSubset # or Mesh for TCP configuration
+       name: backend
+       mesh: example
+     default: # you can only define one: http or tcp
+       http:
+         requests: 5
+         interval: 10s
+         onRateLimit:
+           status: 423
+           headers:
+             - key: "x-kuma-rate-limited"
+               value: "true"
+               append: true
+       tcp:
+         connections: 100
+         interval: 10s
+```
+ 
+### Considered Options
+ 
+#### TCP rate limit for a specific origin
+ 
+Configuration of connection limits is done on listeners. In case of matching from which service a request arrived, we
+need to get some information about the request. There are two options:
+ 
+* use SourceIP filter matching
+* use SNI filter matching
+ 
+##### Use SourceIP matching
+ 
+In this case, we need to configure a listener filter matcher with a list of all possible IPs from which requests can arrive. In the case of one service, it can be many IPs. Matching on many ips might not be the most efficient way and each change of IP requires listener reload.
+ 
+##### Use SNI filter matching
+ 
+It requires a change of SNI that is set on cluster `service_name{mesh=name}` to something different `service_name{mesh=name}{origin=requesting_service}` to later match it in the filter chain. That is not the best idea and can cause issues in multi-zone setups. Apart from that, it requires mTLS to be enabled.
```

Comment:
Having feature being unavailable because of mTLS not being on is not new and I think it's acceptable.
So the only way to do per connection origin is by using SNI? No way to get info from the SAN?

Reply:
Unfortunately, I haven't seen any way to do this with SAN. Filter matcher can only get it by source/destination/protocol/SNI. So there is no way to retrieve SAN.